components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java

```diff
@@ -180,6 +182,45 @@ Get ancestor organization ids (including itself) of a given organization. The li
     }
 
     @Override
+    public void clearBrandingResolverCacheHierarchy(String currentTenantDomain) throws BrandingPreferenceMgtException {
+        String organizationId = getOrganizationId();
+
+        try {
+            OrganizationManager organizationManager =
+                        BrandingResolverComponentDataHolder.getInstance().getOrganizationManager();
+
+            if (organizationId == null) {
+                // If organization id is not available in the context, try to resolve it from tenant domain
+                organizationId = organizationManager.resolveOrganizationId(currentTenantDomain);
+            }
+
+            List<String> childOrganizationIds = new ArrayList<>();
+            childOrganizationIds.add(organizationId);
+
+            // Clear branding resolver caches by looping (breadth-first) through child organization hierarchy
+            while (!childOrganizationIds.isEmpty()) {
+                // Pop the first child organization Id from the list
+                String childOrganizationId = childOrganizationIds.remove(0);
+                String childTenantDomain = organizationManager.resolveTenantDomain(childOrganizationId);
+
+                BrandedOrgCacheKey brandedOrgCacheKey = new BrandedOrgCacheKey(childOrganizationId);
+                BrandedOrgCacheEntry valueFromCache =
+                        brandedOrgCache.getValueFromCache(brandedOrgCacheKey, childTenantDomain);
+                if (valueFromCache != null) {
+                    // If cache exists, clear the cache
+                    brandedOrgCache.clearCacheEntry(brandedOrgCacheKey, childTenantDomain);
+                }
+
+                // Add Ids of all child organizations of the current (child) organization
+                childOrganizationIds.addAll(organizationManager.getChildOrganizationsIds(childOrganizationId));
```

Comment:
If we check the organization has a branding applied, we can avoid cleaning that sub tree. What is the overhead in checking this, compared to retrieving the child orgs and clearing the cache?

Reply:
- In cases where **none of the child orgs have branding**, checking each one individually before clearing the cache can be resource-intensive since there will be a database call for each and every child org before clearing the cache. This can be quite a big overhead, compared to a scenario where we have to check the cache of each and every child org without checking whether the organization has a branding applied.
- But on the other hand, neglecting to check for sub org branding may lead to **unnecessary cache invalidations** for child orgs that have their own branding.

Reply:
Let's think through if we can do any more improvements to this. But looks good to merge.